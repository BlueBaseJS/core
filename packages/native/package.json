{
  "name": "bluerain-boilerplate-native",
  "version": "0.3.0",
  "description": "Hello BlueRain!",
  "author": "BlueEast Team <code@blueeast.com>",
  "private": true,
  "main": "node_modules/expo/AppEntry.js",
  "scripts": {
    "start": "exp start"
  },
  "dependencies": {
    "@blueeast/bluerain-os": "git://github.com/BlueEastCode/bluerain-os.git#feature/custom_platform",
    "expo": "19.0.1",
    "react": "16.0.0-alpha.12",
    "react-dom": "^16.0.0-alpha.12",
    "react-native": "0.46.4",
    "react-native-windows": "^0.33.0",
    "reactxp": "^0.42.0-rc.25",
    "reactxp-navigation": "1.0.5",
    "reactxp-video": "^0.2.0"
  },
  "devDependencies": {
    "@types/node": "^7.0.12",
    "@types/react": "15.0.39",
    "@types/react-native": "0.46.7",
    "@types/webpack": "^2.2.14",
    "source-map-loader": "^0.1.6"
  }
}
