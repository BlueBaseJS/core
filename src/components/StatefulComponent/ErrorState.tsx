import { BlueRainType,withBlueRain } from '../../index';
import Icon from '../Icon';
import React from 'react';

const ErrorIcon = () => <Icon title="Error" />;

const ErrorState = (props: { bluerain: BlueRainType }) => {

	const BR = props.bluerain;

	const ComponentState = BR.Components.get('ComponentState');

	return (
		<ComponentState
			title="Something broke!"
			description="An unknown error has occured. Please try again later."
			image={ErrorIcon}
		/>
	);
};

export default withBlueRain(ErrorState);
