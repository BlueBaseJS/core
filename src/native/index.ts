export * from './ActivityIndicator';
export * from './Button';
export * from './Image';
export * from './Text';
export * from './View';
