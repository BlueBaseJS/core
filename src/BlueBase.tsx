import { Analytics, Logger } from './api';
import {
	ComponentCollection,
	ComponentRegistry,
	ConfigCollection,
	ConfigRegistry,
	HookNestedCollection,
	HookRegistry,
	PluginCollection,
	PluginRegistry,
	ThemeCollection,
	ThemeRegistry,
} from './registries';
import { BlueBaseProvider } from './Context';
import React from 'react';
import { ThemeProvider } from './themes';
import { getComponent } from './getComponent';
import systemHooks from './hooks';

export interface BootOptions {
	components: ComponentCollection,
	configs: ConfigCollection,
	hooks: HookNestedCollection,
	plugins: PluginCollection,
	themes: ThemeCollection,
	// routes: Plugin[]
}

export class BlueBase {

	// APIs
	public Analytics = new Analytics(this);
	public Logger = new Logger(this);

	// Registries
	public Components = new ComponentRegistry(this);
	public Configs = new ConfigRegistry(this);
	public Hooks = new HookRegistry(this);
	public Plugins = new PluginRegistry(this);
	public Themes = new ThemeRegistry(this);

	// Flags
	public booted = false;

	private bootOptions: BootOptions = {
		components: {},
		configs: {},
		hooks: {},
		plugins: [],
		themes: [],
	};

	public async boot(options?: Partial<BootOptions> & { children?: React.ReactNode }) {

		this.bootOptions = { ...this.bootOptions, ...options };

		await this.Hooks.registerNestedCollection(systemHooks);
		await this.Hooks.registerNestedCollection(this.bootOptions.hooks);

		await this.Hooks.run('bluebase.boot', this.bootOptions);

		const SystemApp = getComponent('SystemApp');

		// Set View
		// const SystemApp = this.Components.resolve('SystemApp');
		// SystemApp = await this.Hooks.run('bluebase.system.app', SystemApp);

		const BlueBaseRoot = () => (
			<BlueBaseProvider value={this}>
				<ThemeProvider>
					<SystemApp>
						{options && options.children}
					</SystemApp>
				</ThemeProvider>
			</BlueBaseProvider>
		);

		this.booted = true;

		return BlueBaseRoot;
	}
}