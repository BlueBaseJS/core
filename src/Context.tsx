import { BlueBase } from './BlueBase';
import { createContext } from 'react';

export const BlueBaseContext: React.Context<BlueBase> = createContext(undefined as any);
export const BlueBaseProvider = BlueBaseContext.Provider;
export const BlueBaseConsumer = BlueBaseContext.Consumer;
