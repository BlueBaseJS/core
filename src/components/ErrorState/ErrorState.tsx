import { BlueBase } from '../../BlueBase';
import { BlueBaseContext } from '../../Context';
import React from 'react';

// TODO: show actual error in dev mode
export class ErrorState extends React.PureComponent {

	static contextType = BlueBaseContext;

	render() {

		// FIXME: remove typecasting, added because current react typings don't seem to support this.context
		const BB: BlueBase = (this as any).context;

		return (
			<BB.Components.ComponentState
				title="Something broke!"
				description="An unknown error has occured. Please try again later."
			/>
		);
	}
}
