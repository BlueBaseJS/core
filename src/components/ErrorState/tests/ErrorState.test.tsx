import * as Component from '../..';
import * as Native from '../../../native';
import { BlueBase } from '../../../BlueBase';
import { BlueBaseProvider } from '../../../Context';
import { ErrorState } from '../ErrorState';
import React from 'react';
import TestRenderer from 'react-test-renderer';

beforeEach(() => {
	jest.resetModules();
});

const BB = new BlueBase();
BB.Components.register('Text', Native.Text);
BB.Components.register('View', Native.View);
BB.Components.register('ComponentState', Component.ComponentState);

describe('ErrorState', () => {
	const ErrorStateWithProvider = (props: any) => (
		<BlueBaseProvider value={BB}>
			<ErrorState {...props} />
		</BlueBaseProvider>
	);

	test(`Snapshot ErrorState`, () => {
		const component = TestRenderer.create(
			<ErrorStateWithProvider />
		);
		const tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});

});
