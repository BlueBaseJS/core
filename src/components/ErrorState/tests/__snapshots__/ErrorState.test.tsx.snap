// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorState Snapshot ErrorState 1`] = `
<View>
  <Text>
    Something broke!
  </Text>
  <Text>
    An unknown error has occured. Please try again later.
  </Text>
</View>
`;
